### PR TITLE
feat: Added net.ipv4.tcp_rmem and net.ipv4.tcp_wmem into safe sysctl list

### LINKED ISSUE #125234: Add net.ipv4.tcp_rmem and net.ipv4.tcp_wmem into safe sysctl list
### What would you like to be added?

`net.ipv4.tcp_rmem` and `net.ipv4.tcp_wmem` have been namespaced since https://github.com/torvalds/linux/commit/356d1833b638bd465672aefeb71def3ab93fc17d (Linux Kernal Version >= 4.15)

It would be helpful to allow config these sysctls for each Pod(Application).

### Why is this needed?

The good performance of some applications depends on these sysctls, for example https://docs.datastax.com/en/cassandra-oss/3.x/cassandra/install/installRecommendSettings.html#TCPsettings.

To handle thousands of concurrent connections used by Cassandra, DataStax recommends these settings to optimize the Linux network stack. We should config these sysctls as the following:
```sh
net.ipv4.tcp_rmem = 4096 87380 16777216
net.ipv4.tcp_wmem = 4096 65536 16777216
```
It would be helpful to be able to configure this for applications without needing to modify the kubelet args.

### PR DESCRIPTION
**Pull request by nikzayn**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds the **net.ipv4.tcp_rmem** and **net.ipv4.tcp_wmem** into safe sysctl list because to handle thousands of concurrent connections used by Cassandra, DataStax recommends these settings to optimize the Linux network stack.

#### Which issue(s) this PR fixes:
[#125234](https://github.com/kubernetes/kubernetes/issues/125234)

Fixes #125234

#### Special notes for your reviewer:
I have updated the test cases and it's running fine. Meanwhile, you can check and comment on this PR to suggest some other changes which I might need to update. One callout, I think that I also need to update the **CheckSysctls** function to support checks for the allowed sysctl list.

#### Does this PR introduce a user-facing change?
NONE

### CODE REVIEW

File: pkg/kubelet/sysctl/safe_sysctls.go [hunk 1]
```diff
@@ -60,6 +60,12 @@ var safeSysctls = []sysctl{
 		name:   "net.ipv4.tcp_keepalive_probes",
 		kernel: utilkernel.TCPKeepAliveProbesNamespacedKernelVersion,
 	},
+	{
+		name: "net.ipv4.tcp_rmem",
```
Comment (mauri870) [inline pkg/kubelet/sysctl/safe_sysctls.go:64]: I believe these sysctls are supported starting from kernel 4.15. Can you add a constant to utilkernel and set the kernel property here, like the other entries above? Thanks.

Comment (mauri870) [reply pkg/kubelet/sysctl/safe_sysctls.go:64]: utilkernel will look something like this

```
// TCPReceiveMemoryNamespacedKernelVersion is the kernel version in which net.ipv4.tcp_rmem was namespaced(netns).
// (ref: https://github.com/torvalds/linux/commit/356d1833b638bd465672aefeb71def3ab93fc17d)
const TCPReceiveMemoryNamespacedKernelVersion = "4.15"

// TCPTransmitMemoryNamespacedKernelVersion is the kernel version in which net.ipv4.tcp_wmem was namespaced(netns).
// (ref: https://github.com/torvalds/linux/commit/356d1833b638bd465672aefeb71def3ab93fc17d)
const TCPTransmitMemoryNamespacedKernelVersion = "4.15"
```

**Comment by mauri870** [comment]
Perhaps this also requires changes in `staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go`?

File: pkg/kubelet/sysctl/safe_sysctls.go (hunk 1 above)
Comment (nikzayn) [reply pkg/kubelet/sysctl/safe_sysctls.go:64]: Sure, I will add it. I know that we need to add it just wanted to be confirmed. Thanks!

**Comment by nikzayn** [comment]
> Perhaps this also requires changes in `staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go`?

Thanks, I will take a look and possibly make the changes accordingly. Thanks for reviewing.

File: staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go [hunk 2]
```diff
@@ -118,6 +124,10 @@ func sysctlsV1Dot29(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) Che
 	return sysctls(podMetadata, podSpec, sysctlsAllowedV1Dot29)
 }
 
+func sysctlsV1Dot30(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
```
Comment (mauri870) [inline staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go:127]: I'm not 100% sure this is correct/needed, I'd like a more experienced reviewer to confirm it.

File: staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go [hunk 3]
```diff
@@ -156,4 +156,39 @@ func init() {
...
+			return []*corev1.Pod{
+				// sysctls with out of allowed name
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.SecurityContext.Sysctls = []corev1.Sysctl{{Name: "othersysctl", Value: "other"}}
+				}),
+			}
+		},
+	}
+	registerFixtureGenerator(
+		fixtureKey{level: api.LevelBaseline, version: api.MajorMinorVersion(1, 29), check: "sysctls"},
```
Comment (mauri870) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go:191]: Given that you are using the `V1Dot30` suffix this should be `1, 30`

Comment (nikzayn) [reply staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go:191]: Oh okay, thanks my bad!!. Fixing it right away

File: staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go (hunk 2 above)
Comment (nikzayn) [reply staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go:127]: Yeah, let's wait for an experienced member. Even I also have doubt for the same.

**Comment by thockin** [comment]
This seems fine to me, but I don't know this code path, so I really want someone on sig-node to own the approval.

@pacoxu ?

File: staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go [hunk 4]
```diff
@@ -104,6 +106,10 @@ var (
 		"net.ipv4.tcp_keepalive_intvl",
 		"net.ipv4.tcp_keepalive_probes",
 	)
+	sysctlsAllowedV1Dot30 = sets.NewString(
```
Comment (pacoxu) [inline staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go:109]:
```suggestion
	sysctlsAllowedV1Dot31 = sets.NewString(
```

File: staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go [hunk 5]
```diff
@@ -118,6 +124,10 @@ func sysctlsV1Dot29(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) Che
 	return sysctls(podMetadata, podSpec, sysctlsAllowedV1Dot29)
 }
 
+func sysctlsV1Dot30(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
+	return sysctls(podMetadata, podSpec, sysctlsAllowedV1Dot30)
```
Comment (pacoxu) [inline staging/src/k8s.io/pod-security-admission/policy/check_sysctls.go:128]:
```suggestion
func sysctlsV1Dot31(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
	return sysctls(podMetadata, podSpec, sysctlsAllowedV1Dot31)
```

File: staging/src/k8s.io/pod-security-admission/policy/check_sysctls_test.go [hunk 6]
```diff
@@ -244,3 +266,62 @@ func TestSysctls_1_29(t *testing.T) {
...
+					Sysctls: []corev1.Sysctl{{Name: "net.ipv4.tcp_wmem", Value: "4096 65536 16777216"}},
+				},
+			}},
+			allowed: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			result := sysctlsV1Dot30(&tc.pod.ObjectMeta, &tc.pod.Spec)
```
Comment (pacoxu) [inline staging/src/k8s.io/pod-security-admission/policy/check_sysctls_test.go:311]:
```suggestion
			result := sysctlsV1Dot31(&tc.pod.ObjectMeta, &tc.pod.Spec)
```

**Review by pacoxu — CHANGES_REQUESTED** [review]
This is target to v1.31 and like https://github.com/kubernetes/kubernetes/pull/118846 the test data needs some update.

File: staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go [hunk 7]
```diff
@@ -156,4 +156,39 @@ func init() {
 		fixtureKey{level: api.LevelBaseline, version: api.MajorMinorVersion(1, 29), check: "sysctls"},
 		fixtureDataV1Dot29,
 	)
+
+	fixtureDataV1Dot30 := fixtureGenerator{
```
Comment (mauri870) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go:160]:
```suggestion
	fixtureDataV1Dot31 := fixtureGenerator{
```

Update other references accordingly.

File: staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go [hunk 8]
```diff
@@ -156,4 +156,39 @@ func init() {
...
+			return []*corev1.Pod{
+				// sysctls with out of allowed name
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.SecurityContext.Sysctls = []corev1.Sysctl{{Name: "othersysctl", Value: "other"}}
+				}),
+			}
+		},
+	}
+	registerFixtureGenerator(
+		fixtureKey{level: api.LevelBaseline, version: api.MajorMinorVersion(1, 30), check: "sysctls"},
```
Comment (mauri870) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_sysctls.go:191]:
```suggestion
		fixtureKey{level: api.LevelBaseline, version: api.MajorMinorVersion(1, 31), check: "sysctls"},
```

**Comment by mauri870** [comment]
Thanks for working on this!

**Comment by thockin** [comment]
Is this PR abandoned?  Does it need someone to take it over?

**Comment by pacoxu** [comment]
https://github.com/kubernetes/kubernetes/pull/127489 is based on @nikzayn's commits here and was merged today.

/close

**Pull request closed without merging**